### DISCUSSION TITLE
OpenStreetMap commercial use

### DISCUSSION THREAD

**Discussion by kaartzz** [Mapsui/Mapsui#1463] [Q&A] [answered]
Hi everyone,

I've integrated Mapsui.Forms into my app and displaying OpenStreetMap using the below code

```csharp
var map = new Mapsui.Map {
	CRS = "EPSG:3857",
	Transformation = new Mapsui.Projection.MinimalTransformation()
};
var tileLayer = Mapsui.Utilities.OpenStreetMap.CreateTileLayer();
map.Layers.Insert(0, tileLayer);
mapView.Map = map;
```
It is all working fine. I read in the FAQs that I need to use my own user agent so I don't face any issues like map not showing. How to set the user agent?

And I would like to clarify one more thing. Do I need to add any license for using the default tile source? Is there any limit on using the tile source? Adding it to a commercial product will not cause any issues right?

**Comment by pauldendulk** [comment, marked as answer]
### About the user-agent

The CreateTileLayer method takes an user-agent as optional argument. Not completely sure if this is already in v3 (it is in master). If not you can just past the code of the OpenStreetMap class into your own project.

```csharp
using System.IO;
using BruTile.Predefined;
using BruTile.Web;

namespace Mapsui.Layers.Tiling
{
    public static class OpenStreetMap
    {
        private static readonly BruTile.Attribution OpenStreetMapAttribution = new(
            "© OpenStreetMap contributors", "https://www.openstreetmap.org/copyright");

        public static TileLayer CreateTileLayer(string? userAgent = null)
        {
            userAgent ??= $"user-agent-of-{Path.GetFileNameWithoutExtension(System.AppDomain.CurrentDomain.FriendlyName)}";

            return new TileLayer(CreateTileSource(userAgent)) { Name = "OpenStreetMap" };
        }

        private static HttpTileSource CreateTileSource(string userAgent)
        {
            return new HttpTileSource(new GlobalSphericalMercator(),
                "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
                new[] { "a", "b", "c" }, name: "OpenStreetMap",
                attribution: OpenStreetMapAttribution, userAgent: userAgent);
        }
    }
}
```

As you can see in that code the default actually creates a user-agent based on the name of your application, so that may be sufficient.

### **About the license**

This should be answered by the osm license. One thing is that it should have attribution which points back to the license. The code sample above would create such attribution. I just now read the full license again and it seems to be more restrictive from the last time I read it (which was admittedly a long time ago). 
- Copyright https://www.openstreetmap.org/copyright/en
- API Usage policy https://operations.osmfoundation.org/policies/api/
- Tile Usage Policy https://operations.osmfoundation.org/policies/tiles/

From that last link:

> Heavy use (e.g. distributing an app that uses tiles from openstreetmap.org) is forbidden without prior permission from the [Operations Working Group](https://wiki.osmfoundation.org/wiki/Operations_Working_Group). 

I was not aware of this.

> Cache Tile downloads locally according to HTTP Expiry Header, alternatively a minimum of 7 days.

This should be in our default implementation.

> Maximum of 2 download threads. (Unmodified web browsers’ download thread limits are acceptable.)

It is 4 threads in our default implementation and it is not easy to change.

**Reply by pauldendulk** [reply]
About:

> Heavy use (e.g. distributing an app that uses tiles from openstreetmap.org) is forbidden without prior permission from the [Operations Working Group](https://wiki.osmfoundation.org/wiki/Operations_Working_Group). See below for alternatives.
Clearly display [license](https://wiki.openstreetmap.org/wiki/License) attribution.

Looking at the alternatives below it lists the url we use at the top of the table as 'free'. So not clear what this mean:

https://wiki.openstreetmap.org/wiki/Tile_servers

I wrote an email to the openstreetmap Operations Working group requesting more guidance.

**Reply by kaartzz** [reply]
Thank you very much for the detailed explanation.
I'm using Mapsui 2.0.3 and this version doesn't have OpenStreetMap.CreateTileLayer() that accepts an argument. So I assume the user-agent is set based on the applications name by default and I need not do anything.

When you hear back from OSM working group please let me know, so that I can decide what needs to be done in my app.
Just to make sure, [OpenStreetMap standard tile](https://tile.openstreetmap.org/${z}/${x}/${y}.png) - This is the tile url that mapsui uses to get the tiles, right?

**Reply by pauldendulk** [reply]
No, in 2.0.3 the user-agent is fixed. So in that case it is better to copy the code above to your own project.

**Reply by janusw** [reply]
Or, even better, update to Mapsui 3.0.x, where custom user agents are supported ;)

**Reply by kaartzz** [reply]
Mapsui 3.0 requires Xamarin.Forms > 5.0. Because of some other dependencies in my project I was not able to upgrade Forms and that's why I'm using 2.0.3

By fixed User-agent do you mean it will be set to 'Path.GetFileNameWithoutExtension(System.AppDomain.CurrentDomain.FriendlyName)' by default or I must definitely add this OpenStreetMap class to my project and set the user-agent myself?

**Reply by pauldendulk** [reply]
No, it is a fixed value. Can't remember which, It would be something like 'mapsui-user-agent". So, everyone with v2 will have that same user-agent. This means that if anyone with v2 abuses the osm service (by mass downloading for instance) and that user agent will get blocked, your app will suffer from it. At the moment the user-agent gets blocked your users will get a blank map, and the only way to fix is is by shipping a new version of the app. So, it is advised to copy the block of code above and use that one.
